File: backend/src/app/auth/totp.py
"""RFC 6238 TOTP — stdlib only (no third-party dependency).

A single 6-digit, 30-second, SHA-1 authenticator code (the universal default that
every authenticator app speaks). Secrets are base32 strings; storage encrypts them
with the Fernet helper. ``verify`` accepts a small clock-drift window and takes an
optional ``at`` so tests are deterministic.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import secrets
import struct
import time
from urllib.parse import quote

DIGITS = 6
PERIOD = 30


def generate_secret(num_bytes: int = 20) -> str:
    """Return a fresh base32 secret (no padding), suitable for an otpauth URI."""
    return base64.b32encode(secrets.token_bytes(num_bytes)).decode("ascii").rstrip("=")


def _hotp(secret_b32: str, counter: int) -> str:
    pad = "=" * (-len(secret_b32) % 8)
    key = base64.b32decode(secret_b32.upper() + pad)
    digest = hmac.new(key, struct.pack(">Q", counter), hashlib.sha1).digest()
    offset = digest[-1] & 0x0F
    truncated = struct.unpack(">I", digest[offset : offset + 4])[0] & 0x7FFFFFFF
    return str(truncated % (10**DIGITS)).zfill(DIGITS)


def verify(secret_b32: str, code: str, *, window: int = 1, at: float | None = None) -> bool:
    """Constant-time check of ``code`` against the secret, ±``window`` time steps."""
    code = (code or "").strip()
    if not code.isdigit() or len(code) != DIGITS:
        return False
    now = time.time() if at is None else at
    counter = int(now // PERIOD)
    for drift in range(-window, window + 1):
        if hmac.compare_digest(_hotp(secret_b32, counter + drift), code):
            return True
    return False


def provisioning_uri(secret_b32: str, account: str, issuer: str) -> str:
    """Build the ``otpauth://`` URI the frontend renders as a QR code."""
    label = quote(f"{issuer}:{account}", safe="")
    params = (
        f"secret={secret_b32}"
        f"&issuer={quote(issuer, safe='')}"
        f"&algorithm=SHA1&digits={DIGITS}&period={PERIOD}"
    )
    return f"otpauth://totp/{label}?{params}"

File: orbit/lib/orbit/auth/totp.ex
defmodule Orbit.Auth.TOTP do
  @moduledoc """
  RFC 6238 TOTP — mirror of backend/src/app/auth/totp.py.

  Single 6-digit, 30-second, SHA-1 code (the universal authenticator default).
  Secrets are unpadded base32 strings, stored Fernet-encrypted in
  `users.totp_secret_enc` — existing enrollments must keep validating, proven
  by cross-language vectors in totp_test.exs.

  `verify/3` accepts a small clock-drift window and an optional `:at`
  timestamp so tests are deterministic.
  """

  import Bitwise

  @digits 6
  @period 30

  @spec generate_secret(pos_integer()) :: String.t()
  def generate_secret(num_bytes \\ 20) do
    num_bytes
    |> :crypto.strong_rand_bytes()
    |> Base.encode32(padding: false)
  end

  @spec verify(String.t(), String.t(), keyword()) :: boolean()
  def verify(secret_b32, code, opts \\ []) do
    code = String.trim(code || "")
    window = Keyword.get(opts, :window, 1)
    at = Keyword.get(opts, :at, System.os_time(:second))

    if valid_code_format?(code) do
      counter = div(trunc(at), @period)

      Enum.any?(-window..window, fn drift ->
        :crypto.hash_equals(hotp(secret_b32, counter + drift), code)
      end)
    else
      false
    end
  end

  @spec provisioning_uri(String.t(), String.t(), String.t()) :: String.t()
  def provisioning_uri(secret_b32, account, issuer) do
    label = URI.encode("#{issuer}:#{account}", &URI.char_unreserved?/1)

    params =
      "secret=#{secret_b32}&issuer=#{URI.encode(issuer, &URI.char_unreserved?/1)}" <>
        "&algorithm=SHA1&digits=#{@digits}&period=#{@period}"

    "otpauth://totp/#{label}?#{params}"
  end

  defp valid_code_format?(code) do
    byte_size(code) == @digits and code =~ ~r/^\d+$/
  end

  defp hotp(secret_b32, counter) do
    key = Base.decode32!(String.upcase(secret_b32), padding: false)
    digest = :crypto.mac(:hmac, :sha, key, <<counter::unsigned-big-64>>)
    offset = :binary.last(digest) &&& 0x0F
    <<_::binary-size(^offset), truncated::unsigned-big-32, _::binary>> = digest

    (truncated &&& 0x7FFFFFFF)
    |> rem(Integer.pow(10, @digits))
    |> Integer.to_string()
    |> String.pad_leading(@digits, "0")
  end
end
